fuel-tx/src/transaction/types/upgrade.rs

```diff
@@ -0,0 +1,317 @@
+use crate::{
+    transaction::{
+        id::PrepareSign,
+        metadata::CommonMetadata,
+        types::chargeable_transaction::{
+            ChargeableMetadata,
+            ChargeableTransaction,
+            UniqueFormatValidityChecks,
+        },
+        Chargeable,
+    },
+    ConsensusParameters,
+    GasCosts,
+    Input,
+    Output,
+    TransactionRepr,
+    ValidityError,
+};
+use derivative::Derivative;
+use fuel_types::{
+    bytes::WORD_SIZE,
+    canonical::Serialize,
+    Bytes32,
+    ChainId,
+    Word,
+};
+
+use fuel_crypto::Hasher;
+
+#[cfg(feature = "alloc")]
+use alloc::boxed::Box;
+
+pub type Upgrade = ChargeableTransaction<UpgradeBody, UpgradeMetadata>;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+pub enum UpgradeMetadata {
+    /// The metadata for the upgrade transaction that changes the consensus parameters.
+    ConsensusParameters {
+        /// Deserialized consensus parameters from the witness.
+        consensus_parameters: Box<ConsensusParameters>,
+        /// The actual checksum of the serialized consensus parameters.
+        calculated_checksum: Bytes32,
+    },
+    /// Currently there is no metadata for state transition upgrades, so leave it empty.
+    #[default]
+    StateTransition,
+}
+
+impl UpgradeMetadata {
+    pub fn compute(tx: &Upgrade) -> Result<Self, ValidityError> {
```

Comment:
If the `Upgrade` (`ChargeableTransaction`) already has `Some` metadata, would it make sense to use the "cached" value ever?

Reply:
No, this function calculates new metadata from the `Upgrade` transaction. We shouldn't use old metadata since it may not be actual anymore because of changes that we made.